### PR TITLE
added first spectral processing routines, i.e. radar moment calculation

### PR DESCRIPTION
**Pull request by KarlJohnsonnn**
spcutil.py contains the first processing routines for RPG cloud radar Doppler spectra. 
Below a minimum example for calculating radar moments (Ze, mdv, sw, skew, kurt) from LV0 binaries.

from rpgpy import read_rpg
from rpgpy.spcutil import spectra2moments

header, data = read_rpg('FILENAME.LV0')
moments = spectra2moments(data, header)
print(moments)

### CODE REVIEW

File: rpgpy/spcutil.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    for ispec in range(imaxima, -1, -1):
+        if signal[ispec] > threshold: continue
+        index_left = ispec + 1  # the +1 is important, otherwise a fill_value will corrupt the numba code
+        break
+
+    return threshold, [index_left, index_right]
+
+
+
+def replace_fill_value(data: np.array, newfill: np.array) -> np.array:
```
Comment (tukiains) [inline rpgpy/spcutil.py:129]: Do we need this function?

File: rpgpy/spcutil.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    sw = np.sqrt(np.abs(np.sum(pwr_nrm * vel_diff2)))
+    sw2 = sw * sw
+    skew = np.sum(pwr_nrm * vel_diff * vel_diff2 / (sw * sw2))
+    kurt = np.sum(pwr_nrm * vel_diff2 * vel_diff2 / (sw2 * sw2))
+
+    return np.array((Ze_lin, VEL, sw, skew, kurt), dtype=float())
+
+
+@jit(nopython=True, fastmath=True)
+def find_peak_edges(signal: np.array, threshold: float = -1, imaxima: int = -1) -> Tuple[float, List[int]]:
```
Comment (tukiains) [inline rpgpy/spcutil.py:99]: Do we need these optional arguments?

You should write unit test for this function.

File: rpgpy/spcutil.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    moments = {var: moments[:, :, i] for i, var in enumerate(['Ze', 'MeanVel', 'SpecWidth', 'Skewn', 'Kurt'])}
+
+    for mom in moments.keys():
+        moments[mom][invalid_mask] = -999.0
+
+    return moments
+
+
+@jit(nopython=True, fastmath=True)
```
Comment (tukiains) [inline rpgpy/spcutil.py:59]: Good use of `numba` here.

File: rpgpy/spcutil.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    Note:
+        Divide the signal_sum by 2 because vertical and horizontal channel are added.
+        Subtract half of of the Doppler resolution from mean Doppler velocity, because
+
+    Args:
+        - signal: detected signal from a Doppler spectrum
+        - vel_bins: extracted velocity bins of the signal (same length as signal)
+
+    Returns:
+        dict containing
```
Comment (tukiains) [inline rpgpy/spcutil.py:74]: You are returning an array, not `dict`.

File: rpgpy/spcutil.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+@jit(nopython=True, fastmath=True)
+def radar_moment_calculation(signal: np.array, vel_bins: np.array) -> np.array:
+    """
+    Calculation of radar moments: reflectivity, mean Doppler velocity, spectral width,
+        skewness, and kurtosis of one Doppler spectrum. Optimized for the use of Numba.
+
+    Note:
+        Divide the signal_sum by 2 because vertical and horizontal channel are added.
+        Subtract half of of the Doppler resolution from mean Doppler velocity, because
```
Comment (tukiains) [inline rpgpy/spcutil.py:67]: because... ?

File: rpgpy/spcutil.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        if signal[ispec] > threshold: continue
+        index_right = ispec
+        break
+
+    for ispec in range(imaxima, -1, -1):
+        if signal[ispec] > threshold: continue
+        index_left = ispec + 1  # the +1 is important, otherwise a fill_value will corrupt the numba code
+        break
+
+    return threshold, [index_left, index_right]
```
Comment (tukiains) [inline rpgpy/spcutil.py:125]: No need to return `threshold` if you don't intent to use it.

File: rpgpy/spcutil.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    mask2D = np.all(mask, axis=2)
+    ranges = np.append(LV0meta['RngOffs'], LV0meta['RAltN'])
+
+    for iC in range(LV0meta['SequN']):
+        Dopp_res = np.mean(np.diff(LV0meta['velocity_vectors'][iC]))
+
+        for iR in range(ranges[iC], ranges[iC + 1]):
+            for iT in range(n_ts):
+                if mask2D[iT, iR]: continue
+                _, (lb, rb) = find_peak_edges(spec_lin[iT, iR, :])
```
Comment (tukiains) [inline rpgpy/spcutil.py:38]: `lb`, `rb` could be `edge_left`, `edge_right`

File: rpgpy/spcutil.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        moments[:, ranges[iC]:ranges[iC + 1], 1] -= Dopp_res / 2.0
+
+    # create the mask where invalid values have been encountered
+    invalid_mask = np.full((n_ts, n_rg), True)
+    invalid_mask[np.where(moments[:, :, 0] > 0.0)] = False
+
+    moments = {var: moments[:, :, i] for i, var in enumerate(['Ze', 'MeanVel', 'SpecWidth', 'Skewn', 'Kurt'])}
+
+    for mom in moments.keys():
+        moments[mom][invalid_mask] = -999.0
```
Comment (tukiains) [inline rpgpy/spcutil.py:54]: Fill value could be given as an optional argument.

File: rpgpy/spcutil.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+        threshold (optional): noise threshold
+        imaxima (optional): index of signal maximum
+
+    Returns:
+        threshold, [index_left, index_right]: noise threshold & indices of signal minimum/maximum velocity
+    """
+    len_sig = len(signal)
+    index_left, index_right = 0, len_sig
+    if threshold < 0: threshold = np.min(signal)
+    if imaxima < 0: imaxima = np.argmax(signal)
```
Comment (tukiains) [inline rpgpy/spcutil.py:113]: Here again, break these lines.

**Review by tukiains — CHANGES_REQUESTED** [review]
Generally a clean and efficient-looking implementation an to important feature. Nice work! And good use of Numba to boost the performance. I'm not sure if we can get any speed-up by using Cython (I can try it at some point if I find time).

You could pay more attention on how you name variables. Avoid short names like `iC` and `lb` that are hard to interpret, and in Python, you should generally use lowercase for variable names. I recommend following PEP8 (https://www.python.org/dev/peps/pep-0008/) as a general style guide.

There are no tests included so it's hard to verify this works as expected. You should write unit test at least for `find_peak_edges`. Can we somehow validate the calculated moments (compare with RPG LV1 for example?).

See also the comments in the code.

File: rpgpy/spcutil.py (hunk 5 above)
Comment (KarlJohnsonnn) [reply rpgpy/spcutil.py:67]: This is based on an email conversation with the radar software developer. Tests showed the division by two is necessary to match the reflectivity values in the LV1. Latest tests showed that this differs for newer RPG radar software versions.

**Comment by KarlJohnsonnn** [comment]
I made the changes but Unit tests are still TBD.

File: rpgpy/spcutil.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    Return:
+        container_dict: dictionary of larda containers, including larda container for Ze, VEL, sw, skew, kurt
+
+    """
+    # initialize variables:
+    n_ts, n_rg = LV0data[spec_var].shape[:2]
+    moments = np.full((n_ts, n_rg, 5), np.nan)
+
+    spec_lin = LV0data[spec_var].copy()
+    mask = spec_lin <= 0.0
```
[the remark this answers is not in the document]
Comment (KarlJohnsonnn) [reply rpgpy/spcutil.py:28]: E.g. if the fill value is -999. 
Do we need to check for different fill values?

File: rpgpy/spcutil.py (hunk 2 above)
Comment (KarlJohnsonnn) [reply rpgpy/spcutil.py]: Handling jit during unit test is not that straight forward, I found some answers using environmental variables to disable during testing. Is this solution ok? I wrote a wrapper function and added 2 test cases.

Comment (tukiains) [reply rpgpy/spcutil.py]: I can take a look at this.

**Review by tukiains — APPROVED** [review]
You can merge this now. Test for the `radar_moment_calculation` can be added later.

**Pull request merged**